### PR TITLE
[PFCWD][MASIC] Update pfcwd tests cases to run on multi ASIC platform

### PR DESCRIPTION
**Pull request by smaheshm**
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR

Update tests under pfcwd/test_xx to run on multi ASIC platform. Each test will run on each frontend ASIC of the multi ASIC platform.

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Approach
#### What is the motivation for this PR?
Enhance pytest to run on multi ASIC platform

#### How did you do it?
Update the setup pfcwd fixture to include frontend ASIC index fixture, and generate test ports based on the ASIC instance. 

#### How did you verify/test it?
```
samaddik@svcstr-server-2:/var/sonic-mgmt/tests$ pytest pfcwd/test_pfc_config.py  --testbed=vms12-t1-lag-1 --inventory=../ansible/str --testbed_file=../ansible/testbed.csv --host-pattern=str-s6000-acs-8 --module-path=../ansible/library --disable_loganalyzer --skip_sanity   --log-level DEBUG                                                                        
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
========================================================================================================================================================================== test session starts ===========================================================================================================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.10.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/sonic-mgmt/tests, inifile: pytest.ini
plugins: html-1.22.1, repeat-0.9.1, forked-1.3.0, xdist-1.28.0, metadata-1.11.0, ansible-2.2.2
collecting ... ['conf-name', 'group-name', 'topo', 'ptf_image_name', 'ptf', 'ptf_ip', 'ptf_ipv6', 'server', 'vm_base', 'dut', 'comment']
collected 8 items                                                                                                                                                                                                                                                                                                                                                        

pfcwd/test_pfc_config.py ........                                                                                                                                             [100%]

================================================================================= warnings summary ==================================================================================
/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.dualtor
    self.import_plugin(import_spec)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
====================================================================== 8 passed, 1 warnings in 257.79 seconds =======================================================================

samaddik@svcstr-server-2:/var/sonic-mgmt/tests$ pytest pfcwd/test_pfcwd_all_port_storm.py pfcwd/test_pfcwd_function.py pfcwd/test_pfcwd_timer_accuracy.py  --testbed=vms12-t1-lag-1 --inventory=../ansible/str --testbed_file=../ansible/testbed.csv --host-pattern=str-s6000-acs-8 --module-path=../ansible/library --disable_loganalyzer --skip_sanity   --log-level DEBUG
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
========================================================================================================================================================================== test session starts ===========================================================================================================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.10.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/sonic-mgmt/tests, inifile: pytest.ini
plugins: html-1.22.1, repeat-0.9.1, forked-1.3.0, xdist-1.28.0, metadata-1.11.0, ansible-2.2.2
collected 4 items                                                                                                                                                                                                                                                                                                                                                        

pfcwd/test_pfcwd_all_port_storm.py .                                                                                                                                                                                                                                                                                                                               [ 25%]
pfcwd/test_pfcwd_function.py ..                                                                                                                                                                                                                                                                                                                                    [ 75%]
pfcwd/test_pfcwd_timer_accuracy.py .                                                                                                                                                                                                                                                                                                                               [100%]

============================================================================================================================================================================ warnings summary ============================================================================================================================================================================
/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.dualtor
    self.import_plugin(import_spec)

pfcwd/test_pfcwd_all_port_storm.py::TestPfcwdAllPortStorm::test_all_port_storm_restore[None]
  /usr/local/lib/python2.7/dist-packages/pytest_ansible-2.2.2-py2.7.egg/pytest_ansible/module_dispatcher/v28.py:70: UserWarning: provided hosts list is empty, only localhost is available
    warnings.warn("provided hosts list is empty, only localhost is available")

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================================================================================================================================================ 4 passed, 2 warnings in 1304.21 seconds =================================================================================================================================================================

```

Multi ASIC

```
samaddik@svcstr-server-2:/var/sonic-mgmt/tests$ pytest pfcwd/test_pfc_config.py pfcwd/test_pfcwd_all_port_storm.py pfcwd/test_pfcwd_function.py pfcwd/test_pfcwd_timer_accuracy.py  --testbed=vmsvc1-t1-nmasic-acs-1 --inventory=../ansible/strsvc  --testbed_file=../ansible/testbed.csv --host-pattern=svcstr-nmasic-acs-1  --module-path=../ansible/library --disable_log
analyzer --skip_sanity  --log-level DEBUG                                                                                                                                                                                                                                                                                                                                 
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
========================================================================================================================================================================== test session starts ===========================================================================================================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.10.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/sonic-mgmt/tests, inifile: pytest.ini
plugins: html-1.22.1, repeat-0.9.1, forked-1.3.0, xdist-1.28.0, metadata-1.11.0, ansible-2.2.2
collected 48 items                                                                                                                                                                                                                                                                                                                                                       

pfcwd/test_pfc_config.py ................................                                                                                                                                                                                                                                                                                                          [ 66%]
pfcwd/test_pfcwd_all_port_storm.py ....                                                                                                                                                                                                                                                                                                                            [ 75%]
pfcwd/test_pfcwd_function.py ........                                                                                                                                                                                                                                                                                                                              [ 91%]
pfcwd/test_pfcwd_timer_accuracy.py ....                                                                                                                                                                                                                                                                                                                              [100%]

============================================================================================================================================================================= warnings summary =============================================================================================================================================================================
/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.dualtor
    self.import_plugin(import_spec)

pfcwd/test_pfcwd_all_port_storm.py::TestPfcwdAllPortStorm::test_all_port_storm_restore[3]
  /usr/local/lib/python2.7/dist-packages/pytest_ansible-2.2.2-py2.7.egg/pytest_ansible/module_dispatcher/v28.py:70: UserWarning: provided hosts list is empty, only localhost is available
    warnings.warn("provided hosts list is empty, only localhost is available")

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================================================================================================================================================= 48 passed, 2 warnings in 3009.51 seconds =================================================================================================================================================================

```

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

### CODE REVIEW

**Comment by smaheshm** [comment]
azp /run

File: tests/common/devices/sonic_asic.py [hunk 1]
```diff
@@ -387,6 +340,53 @@ def run_redis_cmd(self, argv=[]):
...
+            port: Port ID
+            queue_num: Queue
+        Returns:
+            Queue OID
+        """
+        redis_cmd = [
+            "redis-cli", "-n", "2", "HGET", "COUNTERS_QUEUE_NAME_MAP",
+            "{}:{}".format(port, queue_num)
+        ]
+        queue_oid = next(iter(self.run_redis_cmd(redis_cmd)), None)
```
Comment (neethajohn) [inline tests/common/devices/sonic_asic.py:377]: Why do we need an iterator here?

File: ansible/roles/test/files/ptftests/pfc_wd.py [hunk 2]
```diff
@@ -80,6 +80,10 @@ def runTest(self):
                                     ip_ttl=63)
                 masked_exp_pkt = Mask(exp_pkt)
                 masked_exp_pkt.set_do_not_care_scapy(scapy.Ether, "dst")
+                masked_exp_pkt.set_do_not_care_scapy(scapy.Ether, "src")
```
[the remark this answers is not in the document]
Comment (smaheshm) [reply ansible/roles/test/files/ptftests/pfc_wd.py:83]: Source MAC changes in routing.

File: tests/common/devices/sonic.py [hunk 3]
```diff
@@ -1340,3 +1341,68 @@ def get_up_ip_ports(self):
...
+        except RunAnsibleModuleFail:
+            return False
+        return True
+
+    def is_backend_portchannel(self, port_channel):
+        mg_facts = self.minigraph_facts(host = self.hostname)['ansible_facts']
+        ports = mg_facts["minigraph_portchannels"].get(port_channel)
+        # minigraph facts does not have backend portchannel IFs
+        if ports is None:
+            return True
```
[the remark this answers is not in the document]
Comment (smaheshm) [reply tests/common/devices/sonic.py:1383]: as mentioned in the comment, right now mg_facts does not include 'backend portchannel interface'. So if the port channel does not exist in mg_facts then it's assumed that it's a backend portchannel interface.

File: tests/common/devices/sonic_asic.py (hunk 1 above)
Comment (smaheshm) [reply tests/common/devices/sonic_asic.py:377]: if command returns nothing, this gets assigned to None. "test" and "assign" in one statement.

**Review by neethajohn — APPROVED** [review]

**Pull request merged**